pushnotifications-integration-tests/src/androidTest/java/com/example/pushnotificationsintegrationtests/ReportingTest.kt

```diff
@@ -0,0 +1,231 @@
+import android.content.Intent
+import android.os.Bundle
+import android.support.test.InstrumentationRegistry
+import android.support.test.runner.AndroidJUnit4
+import com.firebase.jobdispatcher.*
+import com.pusher.pushnotifications.*
+import com.pusher.pushnotifications.fcm.MessagingService
+import com.pusher.pushnotifications.internal.AppActivityLifecycleCallbacks
+import com.pusher.pushnotifications.internal.InstanceDeviceStateStore
+import com.pusher.pushnotifications.reporting.ReportingJobService
+import com.pusher.pushnotifications.reporting.api.DeliveryEvent
+import org.awaitility.core.ConditionTimeoutException
+import org.awaitility.kotlin.await
+import org.awaitility.kotlin.until
+import org.awaitility.kotlin.untilNotNull
+import org.junit.*
+import org.junit.Assert.*
+import org.junit.runner.RunWith
+import java.io.File
+import java.util.concurrent.TimeUnit
+
+/**
+ * Instrumented test, which will execute on an Android device.
+ *
+ * See [testing documentation](http://d.android.com/tools/testing).
+ */
+@RunWith(AndroidJUnit4::class)
+class ReportingTest {
+  val context = InstrumentationRegistry.getTargetContext()
+  val instanceId = "00000000-1241-08e9-b379-377c32cd1e84"
+  val errolClient = ErrolAPI(instanceId, "http://localhost:8080")
+  val targetCtx: Context = InstrumentationRegistry.getTargetContext()
+
+  fun getStoredDeviceId(): String? {
+    val deviceStateStore = InstanceDeviceStateStore(InstrumentationRegistry.getTargetContext(), instanceId)
+    return deviceStateStore.deviceId
+  }
+
+  companion object {
+    val secretKey = "a-really-long-secret-key-that-ends-with-hunter2"
+
+    val errol = FakeErrol(8080, secretKey)
+
+    @AfterClass
+    @JvmStatic
+    fun shutdownFakeErrol() {
+      errol.stop()
+    }
+  }
+
+  @Before
+  @After
+  fun wipeLocalState() {
+    val deviceStateStore = InstanceDeviceStateStore(InstrumentationRegistry.getTargetContext(), instanceId)
+
+    await.atMost(1, TimeUnit.SECONDS) until {
+      assertTrue(deviceStateStore.clear())
+
+      deviceStateStore.interests.size == 0 && deviceStateStore.deviceId == null
+    }
+
+    File(context.filesDir, "$instanceId.jobqueue").delete()
+  }
+
+  @Before
+  @After
+  fun wipeServerReportedNotifications() {
+    errol.getInstanceEventsStorage(instanceId).clear()
+  }
+
+  private fun assertStoredDeviceIdIsNotNull() {
+    try {
+      await.atMost(DEVICE_REGISTRATION_WAIT_SECS, TimeUnit.SECONDS) untilNotNull {
+        getStoredDeviceId()
+      }
+    } catch (e: ConditionTimeoutException) {
+      // Maybe FCM is complaining in CI, so let's pretend to have a token now
+      MessagingService.onRefreshToken!!("fake-fcm-token")
+
+      await.atMost(DEVICE_REGISTRATION_WAIT_SECS, TimeUnit.SECONDS) untilNotNull {
+        getStoredDeviceId()
+      }
+    }
+  }
+
+  @Test
+  @Ignore("The job dispatcher will schedule the job but it isn't triggered within a reasonable timeframe")
+  fun aNotificationSentFromPusherShouldBeReportedTestBroken() {
```

Comment:
> Not sure if there's any value in adding a broken test which is marked as ignored since day one.

If we don't include this test, one could wonder about why we are not doing a more realistic test and choosing a more hacky way of doing it. They might try to write the test themselves only to realise it's not possible, thus wasting time. By leaving it in, we inform why it's not working (see the `@Ignore` comment) and we leave open the possibility of being fixed later on -- if someone knows how to make it work or if we refactor the internal behaviour.